File: esports-backend/src/main/java/com/esports/controller/ProjectController.java
package com.esports.controller;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

import com.esports.dto.ApiResponse;
import com.esports.dto.ProjectCreateDTO;
import com.esports.dto.ProjectResponseDTO;
import com.esports.service.ProjectService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/projects")
@CrossOrigin(origins = "http://localhost:5173")
public class ProjectController {

    @Autowired
    private ProjectService projectService;

    @PostMapping
    @PreAuthorize("hasRole('CREATOR') or hasRole('ADMIN')")
    public ResponseEntity<?> createProject(@Valid @RequestBody ProjectCreateDTO projectData, Authentication auth) {
        try {
            Map<String, Object> response = projectService.createProject(auth.getName(), projectData);
            return ResponseEntity.ok(response);
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("Service not found"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to create project"));
        }
    }

    @GetMapping("/my-projects")
    public ResponseEntity<?> getMyProjects(Authentication auth) {
        try {
            List<ProjectResponseDTO> projects = projectService.getMyProjects(auth.getName());
            return ResponseEntity.ok(projects);
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to get projects"));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getProject(@PathVariable Long id, Authentication auth) {
        try {
            ProjectResponseDTO project = projectService.getProject(auth.getName(), id);
            return ResponseEntity.ok(project);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(ApiResponse.error("Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to get project"));
        }
    }

    @PutMapping("/{projectId}/payment")
    public ResponseEntity<?> updatePaymentStatus(@PathVariable Long projectId, @RequestBody Map<String, Object> paymentData, Authentication auth) {
        try {
            Map<String, Object> result = projectService.updatePaymentStatus(auth.getName(), projectId, paymentData);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to update payment"));
        }
    }

    @PutMapping("/{projectId}/status")
    public ResponseEntity<?> updateProjectStatus(@PathVariable Long projectId, @RequestBody Map<String, Object> statusData, Authentication auth) {
        try {
            Map<String, Object> result = projectService.updateProjectStatus(auth.getName(), projectId, statusData);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied"));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(Map.of("error", e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to update status"));
        }
    }

    @PutMapping("/{projectId}/milestones")
    public ResponseEntity<?> updateMilestones(@PathVariable Long projectId, @RequestBody Map<String, Object> milestoneData, Authentication auth) {
        try {
            Map<String, Object> result = projectService.updateMilestones(auth.getName(), projectId, milestoneData);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to update milestones"));
        }
    }

    @GetMapping("/{projectId}/messages")
    public ResponseEntity<List<Map<String, Object>>> getMessages(@PathVariable Long projectId, Authentication auth) {
        try {
            List<Map<String, Object>> messages = projectService.getMessages(auth.getName(), projectId);
            return ResponseEntity.ok(messages);
        } catch (SecurityException e) {
            return ResponseEntity.status(403).build();
        } catch (Exception e) {
            return ResponseEntity.status(500).build();
        }
    }

    @PostMapping("/{projectId}/messages")
    public ResponseEntity<Map<String, Object>> sendMessage(@PathVariable Long projectId, @RequestBody Map<String, Object> messageData, Authentication auth) {
        try {
            Map<String, Object> response = projectService.sendMessage(auth.getName(), projectId, messageData);
            return ResponseEntity.ok(response);
        } catch (SecurityException e) {
            return ResponseEntity.status(403).build();
        } catch (Exception e) {
            return ResponseEntity.status(500).build();
        }
    }
    
    @DeleteMapping("/{projectId}/reject")
    public ResponseEntity<?> rejectProject(@PathVariable Long projectId, Authentication auth) {
        try {
            Map<String, Object> result = projectService.rejectProject(auth.getName(), projectId);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied or project cannot be rejected"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to reject project"));
        }
    }
    
    @PostMapping("/{projectId}/files/upload")
    public ResponseEntity<?> uploadProjectFile(@PathVariable Long projectId, @RequestParam("file") MultipartFile file, Authentication auth) {
        try {
            Map<String, Object> result = projectService.uploadProjectFile(auth.getName(), projectId, file);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to upload file"));
        }
    }
    
    @GetMapping("/{projectId}/files")
    public ResponseEntity<?> getProjectFiles(@PathVariable Long projectId, Authentication auth) {
        try {
            List<Map<String, Object>> files = projectService.getProjectFiles(auth.getName(), projectId);
            return ResponseEntity.ok(files);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(Map.of("error", "Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(Map.of("error", "Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(Map.of("error", "Failed to get files"));
        }
    }
}

File: esports-backend-dotnet/Upkart1/Controllers/ProjectController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.ComponentModel.DataAnnotations;
using System.Security.Claims;
using Upkart1.DTO;
using Upkart1.Services;

namespace Upkart1.Controllers
{
    [ApiController]
    [Route("api/projects")]
    public class ProjectController : ControllerBase
    {
        private readonly IProjectService _projectService;

        public ProjectController(IProjectService projectService)
        {
            _projectService = projectService;
        }

        [HttpPost]
        [Authorize(Roles = "CREATOR,ADMIN")]
        public IActionResult CreateProject([FromBody] ProjectCreateDTO projectData)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var response = _projectService.CreateProject(email, projectData);
                return Ok(response);
            }
            catch (ArgumentException ex)
            {
                return BadRequest(ApiResponse<object>.ErrorResponse(ex.Message));
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("Service not found"));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to create project"));
            }
        }

        [HttpGet("my-projects")]
        public IActionResult GetMyProjects()
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var projects = _projectService.GetMyProjects(email);
                return Ok(projects);
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to get projects"));
            }
        }

        [HttpGet("{id}")]
        public IActionResult GetProject(long id)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var project = _projectService.GetProject(email, id);
                return Ok(project);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("Project not found"));
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, ApiResponse<object>.ErrorResponse("Access denied"));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to get project"));
            }
        }

        [HttpPut("{projectId}/payment")]
        public IActionResult UpdatePaymentStatus(long projectId, [FromBody] Dictionary<string, object> paymentData)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = _projectService.UpdatePaymentStatus(email, projectId, paymentData);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new { error = "Access denied" });
            }
            catch (Exception)
            {
                return StatusCode(500, new { error = "Failed to update payment" });
            }
        }

        [HttpPut("{projectId}/status")]
        public IActionResult UpdateProjectStatus(long projectId, [FromBody] Dictionary<string, object> statusData)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = _projectService.UpdateProjectStatus(email, projectId, statusData);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new { error = "Access denied" });
            }
            catch (ArgumentException ex)
            {
                return BadRequest(new { error = ex.Message });
            }
            catch (Exception)
            {
                return StatusCode(500, new { error = "Failed to update status" });
            }
        }

        [HttpPut("{projectId}/milestones")]
        public IActionResult UpdateMilestones(long projectId, [FromBody] Dictionary<string, object> milestoneData)
        {
            try
            {
                Console.WriteLine($"UpdateMilestones called for project {projectId}");
                Console.WriteLine($"Milestone data: {System.Text.Json.JsonSerializer.Serialize(milestoneData)}");
                
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = _projectService.UpdateMilestones(email, projectId, milestoneData);
                return Ok(result);
            }
            catch (KeyNotFoundException ex)
            {
                Console.WriteLine($"KeyNotFoundException: {ex.Message}");
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException ex)
            {
                Console.WriteLine($"UnauthorizedAccessException: {ex.Message}");
                return StatusCode(403, new { error = "Access denied" });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Exception: {ex.Message}");
                Console.WriteLine($"StackTrace: {ex.StackTrace}");
                return StatusCode(500, new { error = "Failed to update milestones", details = ex.Message });
            }
        }

        [HttpGet("{projectId}/messages")]
        public IActionResult GetMessages(long projectId)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var messages = _projectService.GetMessages(email, projectId);
                return Ok(messages);
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403);
            }
            catch (Exception)
            {
                return StatusCode(500);
            }
        }

        [HttpPost("{projectId}/messages")]
        public IActionResult SendMessage(long projectId, [FromBody] Dictionary<string, object> messageData)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var response = _projectService.SendMessage(email, projectId, messageData);
                return Ok(response);
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403);
            }
            catch (Exception)
            {
                return StatusCode(500);
            }
        }

        [HttpDelete("{projectId}/reject")]
        public IActionResult RejectProject(long projectId)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = _projectService.RejectProject(email, projectId);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new { error = "Access denied or project cannot be rejected" });
            }
            catch (Exception)
            {
                return StatusCode(500, new { error = "Failed to reject project" });
            }
        }

        [HttpPost("{projectId}/files/upload")]
        public async Task<IActionResult> UploadProjectFile(long projectId, [FromForm] IFormFile file)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = await _projectService.UploadProjectFile(email, projectId, file);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new { error = "Access denied" });
            }
            catch (Exception)
            {
                return StatusCode(500, new { error = "Failed to upload file" });
            }
        }

        [HttpGet("{projectId}/files")]
        public IActionResult GetProjectFiles(long projectId)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var files = _projectService.GetProjectFiles(email, projectId);
                return Ok(files);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(new { error = "Project not found" });
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, new { error = "Access denied" });
            }
            catch (Exception)
            {
                return StatusCode(500, new { error = "Failed to get files" });
            }
        }
    }
}